src/dev/lupio/lupio_blk.cc

```diff
@@ -0,0 +1,242 @@
+            if (rw)
+                r |= LUPIO_BLK_TYPE;    /* Write command */
+            if (err)
+                r |= LUPIO_BLK_ERRR;    /* Error */
+            DPRINTF(LupioBLK, "Read LUPIO_BLK_STAT: %d\n", r);
+
+            /* Acknowledge IRQ */
+            platform->clearPciInt(LUPIO_BLK_INT_ID);
+            break;
+
+        default:
+            r = -1;
+            panic("Unexpected read to the LupioRTC device at address %d!",
+                    addr);
+            break;
+    }
+    return r;
+}
+
+void
+LupioBLK::lupioBLKWrite(const uint8_t addr, uint64_t val64)
+{
+    uint32_t val = val64;
+
+    switch (addr >> 2) {
+        case LUPIO_BLK_NBLK:
+            nblk = val;
+            DPRINTF(LupioBLK, "Write LUPIO_BLK_NBLK: %d\n", nblk);
+            break;
+        case LUPIO_BLK_BLKA:
+            lba = val;
+            DPRINTF(LupioBLK, "Write LUPIO_BLK_BLKA: %d\n", lba);
+            break;
+        case LUPIO_BLK_MEMA:
+            mem = val;
+            DPRINTF(LupioBLK, "Write LUPIO_BLK_MEMA: %d\n", mem);
+            break;
+
+        case LUPIO_BLK_CTRL:
+            err = false;
+            rw = !!(val & LUPIO_BLK_TYPE);
+
+            /* Perform command */
+            lupioBLKCmd();
+
+            /* Raise IRQ if requested */
+            if (val & LUPIO_BLK_IRQE) {
+                irq = val & LUPIO_BLK_IRQE;
+                platform->postPciInt(LUPIO_BLK_INT_ID);
+            }
+            break;
+
+        default:
+            panic("Unexpected write to the LupioRTC device at address %d!",
+                    addr);
+            break;
+    }
+}
+
+
+void
+LupioBLK::lupioBLKCmd(void)
+{
+    int req_len;
+    uint8_t *req_data = nullptr;
+    int64_t offset;
+
+    // Check parameters
+    if ((uint64_t)lba + nblk > nb_blocks) {
+        DPRINTF(LupioBLK, "Bad LBA range\n");
+    }
+
+    // Perform transfer
+    req_len = nblk * SECTOR_SIZE;
+    req_data = new uint8_t[req_len];
+    offset = (int64_t)lba;
+    if (!rw) {
+        // Read command (block -> mem)
+        for (int i = 0; i < req_len; i+=SECTOR_SIZE){
+            image.read(&req_data[i], offset);
+            offset++;
+        }
+
+        dmaWrite(mem, req_len, &dmaEvent, req_data, 0);
+    } else {
+        // Write command (mem -> block)
+        dmaRead(mem, req_len, &dmaEvent, req_data, 0);
+
+        for (int i = 0; i < req_len; i+=SECTOR_SIZE){
+            image.write(&req_data[i], offset);
+            offset++;
+        }
+    }
+    delete[] req_data;
+    return;
+}
+
+AddrRangeList
+LupioBLK::getAddrRanges() const
+{
+    DPRINTF(LupioBLK, "entering addr range function\n");
+    AddrRangeList ranges;
+    ranges.push_back(RangeSize(pioAddr, pioSize));
+    return ranges;
+}
+
+Tick
+LupioBLK::read(PacketPtr pkt)
+{
+    Addr addr = pkt->getAddr() - pioAddr;
+
+    DPRINTF(LupioBLK,
+        "Read request - addr: %#x, size: %#x\n", addr, pkt->getSize());
+
+    uint64_t read_request = lupioBLKRead(addr);
```

Comment:
uint32_t, uint64_t, int64_t all seem to be a bit randomly used. I would try to be more careful in which one is used. I'm not sure what's "right", but I would take some time to think about it in each case.

Reply:
In QEMU the read/write functions take uint64_t (the function prototypes are imposed). Here, if we're not constrained, then lupio devices are 32-bits device.